### ISSUE TITLE
Make it easier to round-trip custom tags.

### ISSUE THREAD

**Issue by simonbuchan** [eemeli/yaml#274]
At the moment, as far as I can tell it's pretty tricky to get yaml@2.0.0-5 to parse and stringify a document containing custom tags in general, but particularly to preserve the tags (e.g. for parsing, modifying / stitching the data, stringifying).

**Describe the solution you'd like**
Preferably something as simple (to use) as:

```js
const source = `
a: !B c
d: !E [f, g]
`;
const value = yaml.parse(source, { preserveTags: true });
assert(value.a instanceof yaml.Tag);
assert(value.a.tag === "!B");
assert(value.a.value === "c");
assert(value.d.tag === "!E");
assert(value.d.value[0] === "f");
assert(value.d.value[1] === "g");

value.h = new yaml.Tag("!I", "j");

assert(yaml.stringify(value) === `\
a: !B c
d: !E [f, g]
h: !I j`);
```

**Describe alternatives you've considered**
So I did get this working with:

```js
class YAMLTag { // note, *not* a yaml node.
  constructor(value) {
    this.value = value;
  }

  toString(...args) {
    return this.value.toString(...args);
  }
}

class AWSRef extends YAMLTag {
  static tag = "!Ref";
}

class AWSGetAttr extends YAMLTag {
  static tag = "!GetAttr";
}

...

const options = {
  customTags: [AWSRef, AWSGetAttr, ...].map((Tag) => ({
    tag: Tag.tag,
    resolve(value) {
      return new Tag(value);
    },
    identify(value) {
      return value instanceof Tag;
    },
  })),
};

```

But this is pretty messy, needs you to know about all the tags ahead of time, and I can't figure out how to get it to handle collections properly, let alone the ambiguous format AWS uses, where `!GetAttr Foo.Bar` is a shorthand for `!GetAttr [Foo, Bar]`, so you have to support both formats.

**Additional context**
Seems like the latter issue might be related to #246 ?

**Comment by eemeli** [comment]
Have you tried using `YAML.parseDocument()`? It's sort of designed for keeping this kind of data within the structure, while allowing the document contents to be manipulated.

**Comment by simonbuchan** [comment]
That *would* make sense, but it means all the rest of the code had to be aware of yaml, and use methods rather than native object property manipulation. So I started with that then I ended up with a bunch of code based on that, then hit needing to support tags, and came up with a hacky solution that seemed to work in 2.0.0-3 by extending from Node from yaml/types.

None of this is your problem, of course!

I guess really what I'm asking for is a more flexible tags interface, that doesn't have to give a list of tags up front and can handle ambiguous scalar/collection tag content, so I could easily add this myself. I'm not sure what the constraints are here, though: in particular why scalar and collections are treated so separately.
